pkg/volume/gce_pd/attacher.go

```diff
@@ -0,0 +1,233 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package gce_pd
+
+import (
+	"fmt"
+	"os"
+	"path"
+	"path/filepath"
+	"strconv"
+	"time"
+
+	"github.com/golang/glog"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/util/exec"
+	"k8s.io/kubernetes/pkg/util/keymutex"
+	"k8s.io/kubernetes/pkg/util/mount"
+	"k8s.io/kubernetes/pkg/util/sets"
+	"k8s.io/kubernetes/pkg/volume"
+)
+
+type gcePersistentDiskAttacher struct {
+	host volume.VolumeHost
+	spec *volume.Spec
+}
+
+var _ volume.Attacher = &gcePersistentDiskAttacher{}
+
+var _ volume.AttachableVolumePlugin = &gcePersistentDiskPlugin{}
+
+// Singleton key mutex for keeping attach/detach operations for the
+// same PD atomic
+// TODO(swagiaal): Once the Mount/Unmount manager is implemented this
+// will no longer be needed and should be rmeoved
+var attachDetachMutex = keymutex.NewKeyMutex()
+
+func (plugin *gcePersistentDiskPlugin) NewAttacher(spec *volume.Spec, pod *api.Pod) (volume.Attacher, error) {
+	return &gcePersistentDiskAttacher{host: plugin.host,
+		spec: spec,
+	}, nil
+}
+
+func (attacher *gcePersistentDiskAttacher) Attach(hostName string, mounter mount.Interface) error {
+	deviceMountPath := attacher.GetDeviceMountPath()
+	notMountPoint, err := mounter.IsLikelyNotMountPoint(deviceMountPath)
+	if err != nil && !os.IsNotExist(err) {
+		return err
+	}
+	if !notMountPoint {
+		return nil
+	}
+
+	volumeSource, readOnly := getVolumeSource(attacher.spec)
+	pdName := volumeSource.PDName
+
+	// Block execution until any pending detach operations for this PD have completed
+	attachDetachMutex.LockKey(pdName)
+	defer attachDetachMutex.UnlockKey(pdName)
+
+	gceCloud, err := getCloudProvider(attacher.host.GetCloudProvider())
+	if err != nil {
+		return err
+	}
+
+	for numRetries := 0; numRetries < maxRetries; numRetries++ {
+		if numRetries > 0 {
+			glog.Warningf("Retrying attach for GCE PD %q (retry count=%v).", pdName, numRetries)
+		}
+
+		if err = gceCloud.AttachDisk(pdName, hostName, readOnly); err != nil {
+			glog.Errorf("Error attaching PD %q: %+v", pdName, err)
+			time.Sleep(errorSleepDuration)
+			continue
+		}
+
+		return nil
+	}
+
+	return err
+}
+
+func (attacher *gcePersistentDiskAttacher) WaitForAttach(timeout time.Duration) (string, error) {
+	ticker := time.NewTicker(checkSleepDuration)
+	defer ticker.Stop()
+	timer := time.NewTimer(timeout)
+	defer timer.Stop()
+
+	volumeSource, _ := getVolumeSource(attacher.spec)
+	pdName := volumeSource.PDName
+	partition := ""
+	if volumeSource.Partition != 0 {
+		partition = strconv.Itoa(int(volumeSource.Partition))
+	}
+
+	sdBefore, err := filepath.Glob(diskSDPattern)
```

Comment:
If the device is found this hack doesn't matter anyway. It was put in place to ensure the device shows up. I'm pretty sure we can get rid of the hack now. But maybe as a separate deliberate action in another PR, not here.